base/compiler/ssair/inlining.jl

```diff
@@ -1321,26 +1418,30 @@ end
 
 function handle_match!(
     match::MethodMatch, argtypes::Vector{Any}, flag::UInt8, state::InliningState,
-    cases::Vector{InliningCase}, allow_abstract::Bool = false)
+    cases::Vector{InliningCase}, allow_abstract::Bool = false, check_sparams::Bool = false)
     spec_types = match.spec_types
     allow_abstract || isdispatchtuple(spec_types) || return false
-    # we may see duplicated dispatch signatures here when a signature gets widened
-    # during abstract interpretation: for the purpose of inlining, we can just skip
-    # processing this dispatch candidate
-    _any(case->case.sig === spec_types, cases) && return true
-    item = analyze_method!(match, argtypes, flag, state)
+    if check_sparams
+        # we may see duplicated dispatch signatures here when a signature gets widened
+        # during abstract interpretation: for the purpose of inlining, we can just skip
+        # processing this dispatch candidate
+        _any(case->case.sig === spec_types, cases) && return true
+        item = analyze_method!(match, argtypes, flag, state, true)
+    else
+        item = analyze_method!(match, argtypes, flag, state)
+    end
```

Comment:
Why not
```suggestion
    # we may see duplicated dispatch signatures here when a signature gets widened
    # during abstract interpretation: for the purpose of inlining, we can just skip
    # processing this dispatch candidate
    _any(case->case.sig === spec_types, cases) && return true
    item = analyze_method!(match, argtypes, flag, state, check_sparams)
```
? Couldn't see any reason to skip the signature duplication check in the `else` branch.

Reply:
I've refactored this, but in the case where there are unmatched type parameters, we need to further analyze since `spec_types` matching may not tell the whole story: see [this comment](https://github.com/JuliaLang/julia/pull/45062#issuecomment-1110447843), though the other special casing ended up not being needed